File: mesmaeker-java-core/src/main/java/fr/xgouchet/mesmaeker/core/ContractValidator.java
package fr.xgouchet.mesmaeker.core;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Xavier F. Gouchet
 */
@RunWith(Parameterized.class)
public abstract class ContractValidator<T, C extends Contract<T>> {

    public static <T, C extends Contract<T>> List<Object[]> generateParameters(Class<C> contractClass) {
        Method[] declaredMethods = contractClass.getDeclaredMethods();


        return Arrays.stream(declaredMethods)
                .filter(method -> method.isAnnotationPresent(Clause.class))
                .filter(method -> method.getParameterTypes().length == 0)
                .filter(method -> method.getReturnType() == Void.TYPE)
                .map(method -> new Object[]{method.getName()})
                .collect(Collectors.toList());
    }

    private final String contractClause;

    protected ContractValidator(String contractClause) {
        this.contractClause = contractClause;
    }

    @Test
    public void validateClause() throws InvocationTargetException, IllegalAccessException, NoSuchMethodException {
        C contract = instantiateContract();
        T subject = instantiateSubject();

        contract.setImplementationInstance(subject);

        Method method = contract.getClass().getDeclaredMethod(contractClause);

        method.invoke(contract);
    }

    protected abstract C instantiateContract();

    protected abstract T instantiateSubject();
}

File: mesmaeker-core/src/main/java/fr/xgouchet/mesmaeker/core/ContractValidator.kt
package fr.xgouchet.mesmaeker.core

import fr.xgouchet.mesmaeker.annotations.Clause
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import java.lang.reflect.InvocationTargetException

/**
 * @author Xavier F. Gouchet
 */
@Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
@RunWith(Parameterized::class)
abstract class ContractValidator<T, out C : BaseContract<T>>
protected constructor(private val contractClause: String) {

    @Test
    @Throws(InvocationTargetException::class, IllegalAccessException::class, NoSuchMethodException::class)
    fun validateClause() {
        val contract = instantiateContract()
        val subject = instantiateSubject()

        contract.withInstance(subject)

        val parameters = contract.generateParameters(contractClause) ?: emptyArray()
        val method = contract.javaClass.methods.firstOrNull {
            (it.name == contractClause) && parametersMatch(parameters, it.parameterTypes)
        }

        if (method == null) {
            throw IllegalStateException("Couldn't match claus $contractClause with the given parameters")
        } else {
            method.invoke(contract, *parameters)
        }
    }

    private fun parametersMatch(parameters: Array<Any?>, parameterTypes: Array<out Class<*>>): Boolean {
        if (parameters.size != parameterTypes.size) {
            return false
        }

        for (i in 0 until parameters.size) {
            val param = parameters[i] ?: continue
            val expectedType = parameterTypes[i]

            val matches = when (expectedType) {
                Integer.TYPE -> (expectedType == Integer.TYPE) || (expectedType == Integer::class.java)
                else -> expectedType.isAssignableFrom(param.javaClass)
            }

            if (!matches) return false
        }

        return true
    }

    protected abstract fun instantiateContract(): C

    protected abstract fun instantiateSubject(): T

    companion object {

        @JvmStatic
        fun <T, C : BaseContract<T>> generateParameters(contractClass: Class<C>): List<Array<Any>> {
            val declaredMethods = contractClass.declaredMethods

            val annotatedMethods = declaredMethods
                    .filter { method -> method.isAnnotationPresent(Clause::class.java) }

            val firstNonVoid = annotatedMethods.firstOrNull { method -> method.returnType != Void.TYPE }
            if (firstNonVoid != null) {
                throw IllegalStateException("Methods annotated with @Clause must return void! Method ${firstNonVoid.name} in ${contractClass.simpleName} doesn't match this requirement.")
            }

            return annotatedMethods
                    .filter { method -> method.returnType == Void.TYPE }
                    .map { method -> arrayOf<Any>(method.name) }
                    .toList()
        }
    }
}
